lib/HTFeed/BackupExpiration.pm

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/perl
+package HTFeed::BackupExpiration;
+
+use strict;
+use warnings;
+use HTFeed::Config qw(get_config);
+use HTFeed::DBTools qw(get_dbh);
+use HTFeed::Volume;
+use Carp;
+use Log::Log4perl qw(get_logger);
+use File::Temp;
+
+sub new {
+  my $class = shift;
+
+  my $self = {
+    storage_name => undef,
+    @_
+  };
+
+  unless ($self->{storage_name}) {
+    croak "$class cannot be constructed without a storage name";
+  }
+
+  bless($self, $class);
+  return $self;
+}
+
+sub run {
+  my $self = shift;
+
+  my $config = get_config('storage_classes');
+  my $storage_config = $config->{$self->{storage_name}};
+  # Get a list of all volumes with this storage name that have one or more old
+  # versions and one or more new versions. This query is about finding sets of
+  # volumes, nothing more granular.
+  my $sth = get_dbh()->prepare(<<'SQL');
+    SELECT namespace,id,MIN(version) AS min_version,MAX(version) AS max_version
+    FROM feed_backups
+    WHERE deleted IS NULL
+      AND storage_name=?
+    GROUP BY namespace,id
+    HAVING COUNT(*) > 1
+      AND min_version < DATE_FORMAT(DATE_SUB(NOW(), INTERVAL 180 DAY),"%Y%m%d%H%i%S")
+      AND max_version > DATE_FORMAT(DATE_SUB(NOW(), INTERVAL 180 DAY),"%Y%m%d%H%i%S")
```

Comment:
See comment on tests below, but I'm not sure this will work as expected if we (say) ingest the item twice on the same day, or if for whatever reason we go for a while without running this expiration script. The desired behavior is that if it has two versions that are both older than 6 months, only one is retained.

Reply:
So I think we want to check for:

* at least two versions
* at least one of them is older than 6 months

i.e. I think we can drop the `min_version` clause